src/python/gudhi/reader_utils.cc

```diff
@@ -0,0 +1,103 @@
+ */
+
+#include <gudhi/Points_off_io.h>
+#include <gudhi/reader_utils.h>
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <pybind11/numpy.h>
+
+#include <vector>
+#include <array>
+#include <string>
+
+namespace py = pybind11;
+
+
+PYBIND11_MODULE(reader_utils, m) {
+      m.attr("__license__") = "MIT";
+      m.def("read_points_from_off_file", [](const std::string& off_file) {
+            Gudhi::Points_off_reader<std::vector<double>> off_reader(off_file);
+            return off_reader.get_point_cloud();
+          },
+        py::arg("off_file"),
+        R"pbdoc(
+    Read points from OFF file.
+
+    :param off_file: An OFF file style name.
+    :type off_file: string
+
+    :returns:  The point set.
+    :rtype: List[List[float]]
+        )pbdoc"
+      );
+      m.def("read_lower_triangular_matrix_from_csv_file", &Gudhi::read_lower_triangular_matrix_from_csv_file<double>,
+        py::arg("csv_file"), py::arg("separator") = static_cast<const char>(';'),
+        R"pbdoc(
+    Read lower triangular matrix from a CSV style file.
+
+    :param csv_file: A CSV file style name.
+    :type csv_file: string
+    :param separator: The value separator in the CSV file. Default value is ';'
+    :type separator: char
+
+    :returns:  The lower triangular matrix.
+    :rtype: List[List[float]]
+        )pbdoc"
+      );
+      m.def("read_persistence_intervals_grouped_by_dimension", &Gudhi::read_persistence_intervals_grouped_by_dimension,
+        py::arg("persistence_file"),
+        R"pbdoc(
+    Reads a file containing persistence intervals.
+    Each line might contain 2, 3 or 4 values: [[field] dimension] birth death
+    The return value is a `dict(dim, list(tuple(birth, death)))`
+    where `dim` is an `int`, `birth` a `float`, and `death` a `float`.
+    Note: the function does not check that birth <= death.
+
+    :param persistence_file: A persistence file style name.
+    :type persistence_file: string
+
+    :returns:  The persistence pairs grouped by dimension.
+    :rtype: Dict[int, List[Tuple[float, float]]]
+        )pbdoc"
+      );
+      m.def("read_persistence_intervals_in_dimension", [](const std::string& persistence_file, int only_this_dim) {
+            // std::vector<std::pair<double, double>> as in read_persistence_intervals_in_dimension
+            // was not giving the good results because of memory alignment
+            // std::vector is mandatory as we don't know in advance the number of persistence intervals to be read
+            std::vector<std::array<double,2>> ret;
+            Gudhi::read_persistence_intervals_and_dimension(persistence_file,
+              boost::make_function_output_iterator([only_this_dim, &ret](std::tuple<int, double, double> t) {
+                if (only_this_dim == get<0>(t) || only_this_dim == -1) ret.push_back({get<1>(t), get<2>(t)});
+              }));
+            return py::array_t<double>(std::vector<std::ptrdiff_t>{static_cast<std::ptrdiff_t>(ret.size()), 2},
```

Comment:
This copies the vector. It is probably ok in this case, but a comment saying that there is a copy would be nice.
I tried the following
```c++
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>
#include <pybind11/stl_bind.h>
#include <pybind11/numpy.h>

namespace py=pybind11;
typedef std::vector<std::array<double,2>> V;
PYBIND11_MAKE_OPAQUE(V);
PYBIND11_MODULE(marc, m) {
  py::bind_vector<V>(m, "VectorPair", py::buffer_protocol());
  m.def("fff",[](){
      V a(400000000, {1,2});
      return py::array(py::cast(std::move(a)));
      });
}
```
I limited virtual memory to 7G, and managed to call the function from python. If there was a copy, I expect that at some point it would require 2 buffers of 400000000*2 double, more than 12G.

Reply:
Ok, now it is quite clear to me how to do it. I exposed the C++ persistence intervals vector with your technique on f03fadc2
The syntax is also clearer and more C++ like. Thanks !